Thread cancellation tokens down to Merlin queue to improve responsiveness.

pittsw opened the pull request:
Some editors (Nuclide for sure, not sure about Atom or VS Code plugins) will send a definition request for every (column, row) pair that the mouse moves over if Ctrl/Cmd is held while moving the mouse (for goto declaration).  In larger files, where Merlin can take several seconds to respond, this can lead to the queue growing to sizes large enough that it will take several minutes to clear.  Since these events are sent by mouse movement, it's likely that stale definition requests in the queue could be asking about an identifier that the mouse isn't still hovering over, which would make the response useless.  This change adds a mechanism where all definition requests except for the most recent one drop themselves from the queue immediately without calling into Merlin, which keeps the queue from growing too large and causing the language server to become unresponsive.

arxanas commented on src/bin/server/processes/merlin.ts:
```diff
@@ -6,16 +6,25 @@ import * as server from "vscode-languageserver";
     this.queue = async.priorityQueue((task, callback) => {
+      if (task.type !== null && this.shortCircuitMap[task.type]) {
+        if (task.enqueuedAt < this.shortCircuitMap[task.type]) {
+          return callback();
```
In the event that the request in cancelled, we still have to return a response over the LSP. Does this actually do so? (We do this by setting the `RequestCancelled` error code on the response message, as per the protocol: https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#response-message. I'm sure that the VS Code LSP adapter code that's being used in this project has a nicer way to do it.)

pittsw commented:
Oh, hovering over things with the Cmd key held down in Nuclide is a `definition` request, not a `hover` request.  It looks like `definition.ts` already has support for cancelation on line 32, but it's after the expensive part of the call.  I went ahead and moved that check to before we call `await find("ml")`, but it looks like Nuclide isn't sending cancelation requests for this event.  I'll dig a bit deeper, but I will go ahead and switch from custom queue logic to better support for using cancelation tokens :)

freebroccolo commented:
Thanks for taking a look at this.

Vscode does make similar requests on ctrl+hover, but there is a short delay (maybe 100ms?) before the feature is triggered. I haven't noticed any responsiveness problems with that feature in particular but I haven't tested much with very large files and very large projects.

Regarding the cancellation tokens, I went through most of the features and added an earlier check. As long as the cancellations are being triggered by the editor, I would think responsiveness would be okay even if the queue starts to fill up since most of those requests should resolve immediately.

In any case, I am open to making changes to the queueing logic if it turns out we really need it.

pittsw commented on src/bin/server/feature/definition.ts:
```diff
@@ -18,8 +18,10 @@ export default function(
         merlin.Position.fromCode(event.position),
       );
-      const response = await session.merlin.query(request, event.textDocument);
-      if (token.isCancellationRequested) return null;
```
I'm removing this line because I added "canceled" as a possible response class, so cancelation is caught on the line just after the await.

pittsw commented:
Added a new iteration, this time for threading the cancellation token to Merlin.  Checking for cancelation before pushing the request onto the Merlin queue didn't help us, because the request was getting enqueued before Nuclide had time to send the cancellation event, so I added another check inside the consumer.

freebroccolo commented on src/bin/server/processes/merlin.ts:
```diff
@@ -58,6 +66,7 @@ export default class Merlin implements server.Disposable {
     { query }: merlin.Query<I, O>,
+    token: server.CancellationToken | null,
     id?: types.TextDocumentIdentifier,
     priority: number = 0,
```
Can we move `token` after `id` and make it optional (`token?`) as well?

arxanas replied:
I think it's better to add it as a non-optional but nullable value, because we don't want the callers to accidentally be able to omit a cancellation token. It should be a deliberate act, otherwise we run the risk of silently degrading performance in a hard-to-diagnose way.

arxanas commented:
@freebroccolo: Unfortunately, with the size of our codebase, most requests take quite a while. For example, getting definitions or hover types each take ~3 seconds. These things very easily add up.

On top of that, getting the errors for the current document can take a few seconds as well. If we keep putting entries in the queue to get errors for old versions of the document, we'll quickly fall behind because we haven't canceled any of those entries.

freebroccolo commented:
Yeah, it makes sense. Threading the cancellation tokens through the queue like this is a good idea given and a relatively minor change given the potential responsiveness improvements. Happy to get this merged when it's ready.
